### PR TITLE
[builder] Add builder module -- formerly ci/build.py

### PR DESCRIPTION
**Pull request by jigold**
@cseed FYI, since you wrote all of this code.

### CODE REVIEW

File: ci/Makefile [hunk 1]
```diff
@@ -86,13 +92,13 @@ restart-proxy:
...
 
-# Paradox: can't push locally, but ci needs ci-utils.  Just use the latest.
+# Paradox: can't push locally, but ci needs builder_utils.  Just use the latest.
 run-local: PROXY_IP=$(shell gcloud compute instances list --filter=name=dk-test --format="value(EXTERNAL_IP)")
 run-local: restart-proxy
 	HAIL_GCP_PROJECT=$(PROJECT) \
 	  HAIL_DOMAIN=$(DOMAIN) \
 	  HAIL_IP=$(IP) \
-	  HAIL_CI_UTILS_IMAGE=gcr.io/hail-vdc/ci-utils:f477fdaa98d0e640f1bc2003ccfe1d2a86cf2d7531ff595476ca2412d28fc678 \
+	  HAIL_BUILDER_UTILS_IMAGE=gcr.io/hail-vdc/builder:f477fdaa98d0e640f1bc2003ccfe1d2a86cf2d7531ff595476ca2412d28fc678 \  ## FIXME: need to get this!
```
Comment (jigold) [inline ci/Makefile]: I tried to get the hash for the builder docker image I made, but google was down. I think we can worry about that later

Comment (akotlar) [reply ci/Makefile]: prefer an issue to fixme

File: ci/ci/github.py [hunk 2]
```diff
@@ -119,7 +119,7 @@ def __init__(self, exception, attributes=None):
 STACKED_PR = 'stacked PR'
 
 
-class PR(Code):
+class PR(builder.Code):
```
Comment (akotlar) [inline ci/ci/github.py]: This can be just Code

File: ci/ci/github.py [hunk 3]
```diff
@@ -294,7 +294,7 @@ def github_status(self):
             self.sha = sha_out.decode('utf-8').strip()
 
             with open(f'{repo_dir}/build.yaml', 'r') as f:
-                config = BuildConfiguration(self, f.read(), deploy=False)
+                config = builder.BuildConfiguration(self, f.read(), deploy=False)
```
Comment (akotlar) [inline ci/ci/github.py]: similarly, still BuildConfiguration

File: ci/ci/github.py [hunk 4]
```diff
@@ -426,7 +426,7 @@ def checkout_script(self):
 '''
 
 
-class WatchedBranch(Code):
+class WatchedBranch(builder.Code):
```
Comment (akotlar) [inline ci/ci/github.py]: same

File: ci/ci/github.py [hunk 5]
```diff
@@ -669,7 +669,7 @@ def config(self):
 (cd {shq(repo_dir)}; {self.checkout_script()})
 ''')
             with open(f'{repo_dir}/build.yaml', 'r') as f:
-                config = BuildConfiguration(self, f.read(), deploy=True)
+                config = builder.BuildConfiguration(self, f.read(), deploy=True)
```
Comment (akotlar) [inline ci/ci/github.py]: same

File: ci/Makefile (hunk 1 above)
Comment (jigold) [reply ci/Makefile]: Agree, but in this case, we shouldn't merge it until I figure this out. However, I wanted feedback from ci in the meantime

File: builder/builder/__init__.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+from .build import BuildConfiguration
+
+__all__ = ['BuildConfiguration']
```
Comment (akotlar) [inline builder/builder/__init__.py]: I think you want to export Code here as well

File: ci/Makefile [hunk 7]
```diff
@@ -24,28 +24,30 @@ ifeq ($(IN_HAIL_CI),1)
 
 PROJECT := $(shell gcloud config get-value project)
 
-CI_UTILS_LATEST = gcr.io/$(PROJECT)/ci-utils:latest
-CI_UTILS_IMAGE = gcr.io/$(PROJECT)/ci-utils:$(shell docker images -q --no-trunc ci-utils | sed -e 's,[^:]*:,,')
+BUILDER_UTILS_LATEST = gcr.io/$(PROJECT)/builder:latest
+BUILDER_UTILS_IMAGE = gcr.io/$(PROJECT)/builder:$(shell docker images -q --no-trunc builder | sed -e 's,[^:]*:,,')
```
Comment (akotlar) [inline ci/Makefile]: I feel like this isn't needed. You already define these in the builder makefile, exactly

https://github.com/hail-is/hail/blob/ee5f23ae2ca8ac69cc328ce8f0d707f944764bff/ci/Makefile#L27

File: ci/Makefile [hunk 8]
```diff
@@ -24,28 +24,30 @@ ifeq ($(IN_HAIL_CI),1)
...
 
-build-ci-utils:
-	make -C ../docker build
-	-docker pull $(CI_UTILS_LATEST)
-	python3 jinja2_render.py '{"base_image":{"image":"base"}}' Dockerfile.ci-utils Dockerfile.ci-utils.out
-	docker build -t ci-utils -f Dockerfile.ci-utils.out --cache-from ci-utils,$(CI_UTILS_LATEST),base .
+build-builder-utils:
+	make -C ../builder \
+	  BUILDER_UTILS_LATEST="$(BUILDER_UTILS_LATEST)" \
+          BUILDER_UTILS_IMAGE="$(BUILDER_UTILS_IMAGE)" \
```
Comment (akotlar) [inline ci/Makefile]: Similarly, if we don't define these, drop them and use the builder ones.

File: ci/Makefile [hunk 9]
```diff
@@ -24,28 +24,30 @@ ifeq ($(IN_HAIL_CI),1)
...
 
-push-ci-utils: build-ci-utils
-	docker tag ci-utils $(CI_UTILS_LATEST)
-	docker push $(CI_UTILS_LATEST)
-	docker tag ci-utils $(CI_UTILS_IMAGE)
-	docker push $(CI_UTILS_IMAGE)
+push-builder-utils: build-builder-utils
+	make -C ../builder \
+	  BUILDER_UTILS_LATEST="$(BUILDER_UTILS_LATEST)" \
+          BUILDER_UTILS_IMAGE="$(BUILDER_UTILS_IMAGE)" \
```
Comment (akotlar) [inline ci/Makefile]: same

File: ci/Makefile [hunk 10]
```diff
@@ -54,17 +56,21 @@ push: build
...
+	python3 ../builder/jinja2_render.py '{"code":{"sha":"$(shell git rev-parse --short=12 HEAD)"},"deploy":true,"default_ns":{"name":"default"},"ci_image":{"image":"$(CI_IMAGE)"},"global":{"project":"$(PROJECT)","ip":"$(IP)","domain":"$(DOMAIN)"},"builder_utils_image":{"image":"$(BUILDER_UTILS_IMAGE)"},"ci_database":{"user_secret_name":"sql-ci-ci-user-config"}}' deployment.yaml deployment.yaml.out
 	kubectl -n default apply -f deployment.yaml.out
 else
-build-ci-utils:
+build-builder-utils:
 	make -C ../docker build
-	docker build -t ci-utils -f Dockerfile.ci-utils .
+	make -C ../builder \
+	  BUILDER_UTILS_LATEST="$(BUILDER_UTILS_LATEST)" \
+          BUILDER_UTILS_IMAGE="$(BUILDER_UTILS_IMAGE)" \
```
Comment (akotlar) [inline ci/Makefile]: same

File: builder/Makefile [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+push: build
+	docker tag builder $(BUILDER_UTILS_LATEST)
+	docker push $(BUILDER_UTILS_LATEST)
+	docker tag builder $(BUILDER_UTILS_IMAGE)
+	docker push $(BUILDER_UTILS_IMAGE)
+else
+build:
+	make -C ../docker build
+	python3 jinja2_render.py '{"base_image":{"image":"base"}}' Dockerfile Dockerfile.out
```
Comment (akotlar) [inline builder/Makefile]: Do you want a -docker pull `$(CI_LATEST)` here for caching purposes?

File: builder/Makefile [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+	docker build -t builder -f Dockerfile.out .
```
Comment (akotlar) [inline builder/Makefile]: Similarly, cache-from, [as seen here](https://github.com/hail-is/hail/blob/master/ci/Makefile#L49)

File: builder/Makefile (hunk 11 above)
Comment (akotlar) [reply builder/Makefile]: Ah I see, this is outside IN_HAIL_CI...nvm

File: ci/Makefile (hunk 7 above)
Comment (jigold) [reply ci/Makefile]: I couldn't figure out how to get the variable values back to the ci makefile so they could be used in the jinja template further down.

File: ci/Makefile (hunk 1 above)
Comment (jigold) [reply ci/Makefile]: Fixed

File: ci/Makefile (hunk 7 above)
Comment (akotlar) [reply ci/Makefile]: Ah

**Pull request merged**